SalesforceCommerceCloud/pwa-kit
# @W-15163165@ Reduce session churn / fix SFRA -> PWA session share on private

This PR has an accompanying changeset in plugin_slas: https://github.com/SalesforceCommerceCloud/plugin_slas/pull/187

Currently on hybrid sites, when a shopper navigates between SFRA and PWA, we trigger a SLAS refresh token login to sync the latest session information between the two systems. This refresh token login includes a call to the OCAPI `/sessions` endpoint to session bridge SLAS with ECOM, and results in a new DWSID being generated.

Because of this constant replacing of the DWSID on every transition between systems, code that relies on the DWSID remaining the same are unable to function as expected. In addition, this existing approach relies on tracking changes in the refresh token to trigger these refresh logins; this existing approach does not work for private clients as the refresh token in a private client environment is not discarded after one use.

This PR, in conjunction with changes in plugin_slas, reworks how session information is shared between the two systems so that it is not as reliant on the refresh token or refresh logins.

The new approach has the PWA look out for `cc-at` cookies set by SFRA. If these 'cc-at' cookies, which contain an access token from when SFRA logged into SLAS, exist, the PWA discards its current access token and replaces it with the SFRA token.

This new mechanism works because the the access token provided by SFRA has already gone through session bridging in plugin_slas. 


Testing:
* Deploy this PR and the accompanying plugin_slas changes onto a hybrid environment
* Once your hybrid environment is up, start navigating between PWA and SFRA
* In your dev console, see that after the first visit to SFRA, your DWSID no longer changes unless you log in/log out
* See that any changes you make in PWA (ie. add an item to cart) are reflected in the SFRA session and vice versa (ie. edit a line item quantity in SFRA)
* Log into one system, see that your log in status is shared with the other system
* Log out of one system, see that your logout status is applied to both systems
* Verify that this works on both SLAS public and private clients
* One change you will see is that on every navigation to SFRA after the first, you will no longer be prompted for tracking consent.
* Verify that the session sharing is applied across multiple tabs.
  * Have one tab in PWA, one tab in SFRA. See that change you make in one tab are reflected in the other.
  * For example, change the line item quantity in SFRA or add a product to the cart in PWA
